src/v/utils/log_hist.h

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include <seastar/core/metrics_types.hh>
+
+#include <boost/intrusive/list.hpp>
+
+#include <bit>
+#include <chrono>
+#include <cstdint>
+#include <vector>
+
+/*
+ * A histogram implementation
+ * The buckets upper bounds are powers of 2 minus 1.
+ * `first_bucket_upper_bound` therefore must be a power of 2.
+ * The number of values represented by each bucket increases by powers 2 as
+ * well.
+ *
+ * Assume `number_of_buckets` is 4 and `first_bucket_upper_bound` is 16 the
+ * bucket value ranges are;
+ *
+ * [1, 16), [16, 32), [32, 64), [64, 128)
+ *
+ * And if 1, 16, 32, and 33 are recorded the buckets will have the following
+ * counts;
+ *
+ * [1, 16)   = 1
+ * [16, 32)  = 1
+ * [32, 64)  = 2
+ * [64, 128) = 0
+ */
```

Comment:
Hmm so we always need to have the min bucket start at 1 and the last bucket is also implicitly the "infinity" bucket?

I see this kind of matches the seastar histogram but a bit surprised as it's quite a bit different to prometheus histograms.

Reply:
I think first bucket should be `[0, ...` not `[1, ...` since otherwise where does 0 go? In math `lg(0)` is a ... problem but with lzcnt it is not, it's just 1 less than `lg(1)`.

I think this approach is consistent with prom on the low end? the above would have series

```
le="15"
le="31"
le="63"
```

etc.

Reply:
> I think this approach is consistent with prom on the low end?

Depends on how you look at it I guess. Lets say the range you are interested in is 100 to 200 (some unit) then your smallest bucket would be le="100" yes but I think you can't do something like that with the current approach. 

(The current approach can probably be easily extended to support that by using same base offset but not needed now I guess).

Reply:
> Depends on how you look at it I guess. Lets say the range you are interested in is 100 to 200 (some unit) then your smallest bucket would be le="100" yes but I think you can't do something like that with the current approach.

Just to be clear though the min bucket can be > 1, it just has to be a power of two. So it could be 64 or 128, but not 100 in your example.

At least that's how I understand it.

Reply:
Having the first bucket start at 0 and the last bucket end at inf just simplifies the logic for recording values. Otherwise say we had a non-zero start value `a` for the first bucket and a non-zero end value `b` for the last bucket. The first thing we'd have to do is check the bounds to see if the for a given value `v` that `v in [a, b]`. And if v is outside that range we'd either have to not record that value or clamp that value to be in `[a, b]` then record it. Having the first bucket start a zero and the last bucket end at inf basically does that clamping without any additional instructions. 

That being said I definitely get your point about how things are odd if you're only interested in the `[100, 200]` range and you want a histogram that reflects that. For this histogram implementation the best you could get is two buckets `[0, 128), [128, 256)` of which you'd really only be interested in some of the values of either. My recommendation would be to shift the range to `[0, 100]` with just a subtraction instruction and then you could have a decently granular histogram `[0, 2), [2, 4), [4, 8), [8, 16), ... , [64, 128)`.

That being said this histogram implementation is definitely not designed to be a general use histogram like `hdr_hist` and is instead just designed for our use case of tracking latency.